\section{Datasets}

In this thesis following publically available datasets from scRNAseq experiments were used.

\subsection{PBMC}

This dataset is publically available in 10X Genomics website.
In this experiment, human peripheral blood mononuclear cells (PBMCs) were extracted from fresh whole peripheral blood samples obtained from StemExpress. PBMCs were isolated using SepMate density centrifugation methods.
The library was generated from around 8000 cells (5140 cells recovered) using the Chromium Single Cell 3' v3.1 Reagent Kit,
and sequenced on Illumina NovaSeq 6000 to a read depth of approximately 35000 mean reads per cell.
The transcript reads have length of 90bp.
All this information (and more) is available at
\href{https://www.10xgenomics.com/datasets/5k-human-pbmcs-3-v3-1-chromium-controller-3-1-standard}{10X Genomics website}.

\subsection{Transcriptomic references}

\section{Enhancing transcriptomic reference}

Here is provided general description of the pipeline.

\begin{enumerate}
    \item Map reads with transcriptomic reference.
    \item Take unassigned (and unique) reads.
    \item Split into intersecting and intergenic reads.
    \begin{enumerate}
        \item For intersecting:
        \begin{enumerate}
            \item Cluster.
            \item Filter-out relatively small clusters (custom threshold).
            \item Make IGV snapshots.
            \item Resolve overlapping genes that have some reads.
            \item From the second reference and further: add genes to the original GTF that contain reads and do not overlap with entries from the original.
        \end{enumerate}
        \item For intergenic:
        \begin{enumerate}
            \item Cluster.
            \item Filter-out relatively small clusters (custom threshold).
            \item For the first reference only: filter-out AT-rich reads (clusters?).
            \item For reads that have been left unexplained, repeat from the beginning with the next reference.
            \item For the last reference only: clusters that start just after 3' ends are assigned to genes (i.e., extend genes).
            \item For the last reference only: add largest intergenic unexplained regions to GTF (INTERGENIC entries).
        \end{enumerate}
    \end{enumerate}
    \item Create final GTF and map initial sequences to it.
    \item Compute statistics (reads mapped, genes captured).
    \item Check clustering and other steps (in Jupyter notebooks).
\end{enumerate}
