\section{Data Acquisition}

\paragraph{\textbf{scRNAseq datasets}}

We analyzed datasets from four different tissues: brain, blood, lung, and eye.
All samples, except for two Peripheral Blood Mononuclear Cell (PBMC) datasets, were generated using the 10x Genomics v3.1 protocol.
The two exceptions were prepared using the inDrops2 and inDrops protocols.
All protocols capture short reads from the 3' ends of RNA molecules.
The datasets are publicly available, with sources and additional details provided in Table \ref{tab:datasets}.

\input{"textparts/tables/datasets.tex"}

\paragraph{\textbf{Genome and Transcriptomic references}}

The human genome GRCh38 was used in this project, downloaded from the \href{https://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/dna/}{Ensembl website}.

Four transcriptomic references were analyzed (see Table \ref{tab:references} for details).
To ensure compatibility with the genome FASTA file, chromosome name prefixes ('chr') were removed from references that included them,
as the genome file does not use these prefixes.
This modification was performed using basic Linux command-line tools.
Additionally, for the NCBI reference, chromosome names were converted to Ensembl-style notation using a custom Python script
(e.g., 1, 2, 3 instead of NC\_000001.11, NC\_000002.12, NC\_000003.12, etc.).

For references that do not have 'gene' entries, such entries were added
(entry that spans all the components of a particular gene).

\input{"textparts/tables/references.tex"}

\paragraph{\textbf{Gene Prediction Tracks and Conservation Scores}}

Gene predictions (tracks 'AUGUSTUS', 'Geneid genes', 'Gescan genes', 'SGP genes', 'SIB genes') and 
genome conservation scores ('phastCons100way' track) were downloaded from \href{https://genome.ucsc.edu/}{UCSC genome browser}.

\section{Computational Tools and Environment}

All analyses were performed on a high-performance computing (HPC) cluster running a Linux environment.
Software packages and tools were managed using Conda.
The main tools and their versions are listed in Table \ref{tab:tools}.
The exact Conda environment specifications (YAML file) can be found on [GitHub](GIVE LINK).
The specific functions and parameters used are described in the following sections, where the general analysis pipeline will be explained.

\input{"textparts/tables/tools.tex"}

\section{Enhancing transcriptomic reference}

Here is provided general description of the pipeline.

\begin{enumerate}
    \item Map reads with transcriptomic reference.
    \item Take unassigned (and unique) reads.
    \item Split into intersecting and intergenic reads.
    \begin{enumerate}
        \item For intersecting:
        \begin{enumerate}
            \item Cluster.
            \item Filter-out relatively small clusters (custom threshold).
            \item Make IGV snapshots.
            \item Resolve overlapping genes that have some reads.
            \item From the second reference and further: add genes to the original GTF that contain reads and do not overlap with entries from the original.
        \end{enumerate}
        \item For intergenic:
        \begin{enumerate}
            \item Cluster.
            \item Filter-out relatively small clusters (custom threshold).
            \item For the first reference only: filter-out AT-rich reads (clusters?).
            \item For reads that have been left unexplained, repeat from the beginning with the next reference.
            \item For the last reference only: clusters that start just after 3' ends are assigned to genes (i.e., extend genes).
            \item For the last reference only: add largest intergenic unexplained regions to GTF (INTERGENIC entries).
        \end{enumerate}
    \end{enumerate}
    \item Create final GTF and map initial sequences to it.
    \item Compute statistics (reads mapped, genes captured).
    \item Check clustering and other steps (in Jupyter notebooks).
\end{enumerate}
